Undefined behaviour in integer conversion

Negation of -2147483648 fails when compiled with gcc-4.9 -O2 optimization. Also compiled with -fsanitize=undefined for test log below.

```
==========================================
   QuickFIX 1.14.3: test/test-suite.log
==========================================

# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: check.sh
==============

ut: no process found
at: no process found
<ut>
  <output>
C++/FieldConvertors.h:47:25: runtime error: negation of -2147483648 cannot be represented in type 'int'; cast to an unsigned type to negate this value to itself
C++/FieldConvertors.h:100:25: runtime error: negation of -2147483648 cannot be represented in type 'int'; cast to an unsigned type to negate this value to itself
C++/FieldConvertors.h:190:13: runtime error: negation of -2147483648 cannot be represented in type 'int'; cast to an unsigned type to negate this value to itself
C++/test/FieldConvertorsTestCase.cpp:82: error: Failure in countIntegerSymbols: Expected 11 but was 5
C++/test/FieldConvertorsTestCase.cpp:114: error: Failure in integerConvertTo: Expected -2147483648 but was -00
FAILURE: 2 out of 186 tests failed (2 failures).
Test time: 10.87 seconds.
  </output>
</ut>
```

--- a/src/C++/FieldConvertors.h
+++ b/src/C++/FieldConvertors.h
@@ -37,14 +37,18 @@ namespace FIX
 typedef int signed_int;
 typedef unsigned int unsigned_int;
 
-#define UNSIGNED_VALUE_OF( x ) unsigned_int( x < 0 ? -x : x )
+inline unsigned_int unsigned_value_of( signed int x )
+{
+  auto y = static_cast<unsigned_int>( x );
+  return x < 0 ? -y : y;
+}
 
 #define IS_SPACE( x ) ( x == ' ' )
 #define IS_DIGIT( x ) ( unsigned_int( x - '0' ) < 10 )
 
 inline int number_of_symbols_in( const signed_int value )
 {
-  unsigned_int number = UNSIGNED_VALUE_OF( value );
+  unsigned_int number = unsigned_value_of( value );
 
   int symbols = 0;
 
@@ -97,7 +101,7 @@ inline char* integer_to_string( char* buf, const size_t len, signed_int t )
 
   *--p = '\0';
   
-  unsigned_int number = UNSIGNED_VALUE_OF( t );
+  unsigned_int number = unsigned_value_of( t );
 
   while( number > 99 )
   {

--- a/src/C++/test/FieldConvertorsTestCase.cpp
+++ b/src/C++/test/FieldConvertorsTestCase.cpp
@@ -111,7 +111,7 @@ TEST(integerConvertTo)
   CHECK_EQUAL( "-12345678", IntConvertor::convert( -12345678 ) );
   CHECK_EQUAL( "-123456789", IntConvertor::convert( -123456789 ) );
   CHECK_EQUAL( "-2147483647", IntConvertor::convert( -2147483647 ) );
-  //CHECK_EQUAL( "-2147483648", IntConvertor::convert( MIN_INT ) );
+  CHECK_EQUAL( "-2147483648", IntConvertor::convert( MIN_INT ) );
 }
 
 TEST(integerConvertFrom)

